### PR TITLE
Comments indents

### PR DESCRIPTION
**Pull request by PolinaShneider**
Added comments formatting feature

This will close axibase/charts-editor#17

demo

### CODE REVIEW

File: src/formatter.ts [hunk 1]
```diff
@@ -350,6 +364,62 @@ export class Formatter {
         this.formattedText.push(this.currentIndent + line.trim())
     }
 
+    /**
+     * Handle block comment
+     * @param line 
+     */
+    private handleCommentBlock(line: string): void {
```
Comment (annstriganova) [inline src/formatter.ts]: Could it be wrapped to `LanguageFormatter`? For example, `class CommentsFormatter implements LanguageFormatter`. In this case block comments will be processed in `handleCodeBlock`.

File: src/formatter.ts [hunk 2]
```diff
@@ -350,6 +364,62 @@ export class Formatter {
...
+        const [, before, after] = line.match(/(\/\*+)(.*)/);
```
Comment (annstriganova) [inline src/formatter.ts]: Rename `before` -> `commentStart`.

File: src/formatter.ts [hunk 3]
```diff
@@ -350,6 +364,62 @@ export class Formatter {
...
+        this.indentLine(before);
+        /** Push setting after comment start to line stream */
+        if (!isEmpty(after)) {
+            this.lineStreamPush(after);
+        }
+        line = this.nextLine();
+        while (line != undefined) {
+            if (/(.*)(\*\/)/.test(line)) {
+                // Assuming in is not inline block comment
+                let [, before, after] = line.match(/(.*?)(\*+\/)/);
```
Comment (annstriganova) [inline src/formatter.ts]: Rename `after` -> `commentEnd`.

File: src/formatter.ts [hunk 4]
```diff
@@ -350,6 +364,62 @@ export class Formatter {
...
+
+        /** Write comment start symbol */
+        this.indentLine(before);
+        /** Push setting after comment start to line stream */
+        if (!isEmpty(after)) {
+            this.lineStreamPush(after);
+        }
+        line = this.nextLine();
+        while (line != undefined) {
+            if (/(.*)(\*\/)/.test(line)) {
```
Comment (annstriganova) [inline src/formatter.ts]: It's [better to cache](https://stackoverflow.com/a/32523333/8165307) even literal regular expressions.

File: src/formatter.ts [hunk 5]
```diff
@@ -350,6 +364,62 @@ export class Formatter {
...
+    }
+
+    private dumpBuffer() {
+        this.increaseIndent();
+
+        for (let line of this.commentsBuffer) {
+            this.formattedText.push(
+                this.currentIndent + line.substring(
+                    Math.min(...this.commentBlockIndents)
+                ).trimRight()
```
Comment (annstriganova) [inline src/formatter.ts]: `trimRight` is a non-standard alias, use `trimEnd` instead.

File: src/formatter.ts (hunk 1 above)
Comment (PolinaShneider) [reply src/formatter.ts]: I believe, languages formatting and comments are different for the next reasons.
1) Not sure that block comments could be ideologically considered as a language.
2) Block comment start such as `/* hello` should processed as two lines `/*` and `  hello`. Second line with indent is pushed back to config lines stream. For languages we just have `script | sql | csv` start keyword, and no more text the same line
3) Comments block doesn't need special parsing | tokenizing | formatting, etc. We just preserve formatting as is and add indent.

File: src/formatter.ts [hunk 6]
```diff
@@ -365,59 +379,59 @@ export class Formatter {
...
         }
     }
 
-    private pushBuffer(line: string): void {
-        let indent = line.search(/[^ ]/);
-        if (indent >= 0) {
-            this.commentBlockIndents.push(indent);
+    /** Push line of comment block to buffer and calculate indent */
+    private pushCommentBuffer(line: string): void {
+        const indent = line.search(/[^ ]/);
```
Comment (annstriganova) [inline src/formatter.ts]: Method `pushCommentBuffer` is called in a loop. Each iteration new object for `/[^ ]/` is created.

File: src/formatter.ts [hunk 7]
```diff
@@ -365,59 +379,59 @@ export class Formatter {
...
+        const { lines, minIndent } = this.commentsBuffer;
         this.increaseIndent();
 
-        for (let line of this.commentsBuffer) {
+        for (let line of lines) {
             this.formattedText.push(
-                this.currentIndent + line.substring(
-                    Math.min(...this.commentBlockIndents)
-                ).trimRight()
+                this.currentIndent + line.substring(minIndent).trimEnd()
```
Comment (annstriganova) [inline src/formatter.ts]: Could you explain or just provide an example, why it can not be replaced with

```js
this.currentIndent + line.trim()
```
?

Comment (PolinaShneider) [reply src/formatter.ts]: See unwanted formatting behaviour without preserving spaces. Script contents won't be formatted in commented block

image

File: src/formatter.ts [hunk 8]
```diff
@@ -404,6 +492,14 @@ export class Formatter {
         return this.getLine(++this.currentLine);
     }
 
+    /**
+     * Insert line to the config
+     * @param line 
+     */
+    private lineStreamPush(line: string): void {
+        this.lines.splice(this.currentLine + 1, 0, line);
```
Comment (annstriganova) [inline src/formatter.ts]: For which reason do we need to modify initial config?

Comment (PolinaShneider) [reply src/formatter.ts]: When symbol of comment start and setting are placed the same line
```
/* hello
  test
bye */
```

And we want
```
/* 
  hello
    test
  bye 
*/
```

Comment (PolinaShneider) [reply src/formatter.ts]: So we push setting after `/*` back to config to process it next step

**Pull request merged**